src/js/heuristicblocking.js

```diff
@@ -101,18 +99,15 @@ HeuristicBlocker.prototype = {
       return {};
     }
 
-    let fqdn = (new URI(details.url)).host,
-      origin = window.getBaseDomain(fqdn);
-
-    // if this is a main window request
+    // if this is a main window request, the webrequest listener will handle it
     if (details.type == "main_frame") {
-      // save the origin associated with the tab
-      log("Origin: " + origin + "\tURL: " + details.url);
-      tabOrigins[details.tabId] = origin;
       return {};
     }
 
-    let tabOrigin = tabOrigins[details.tabId];
+    let fqdn = (new URI(details.url)).host,
+      origin = window.getBaseDomain(fqdn),
+      mainFrame = badger.tabData[details.tabId].frames[0],
```

Comment:
Unfortunately `tabData.frames` is unreliable as it doesn't get populated for either inline nor cached frames. ([Previously](https://github.com/EFForg/privacybadger/pull/2167#discussion_r218982802).)

You can see the problem by browsing around and quickly running into `badger.tabData[details.tabId].frames[details.frameId] is undefined` and `badger.tabData[details.tabId] is undefined` errors.